src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java

```diff
@@ -1,156 +1,331 @@
+                { new TestDataEnvelope(getNBytesWithValues(1000, (n, index) -> (byte) index.intValue())) },
+                { new TestDataEnvelope(getNBytesWithValues(1000, (n, index) -> index < n / 2 ? (byte) 0 : (byte) 1)) },
+                { new TestDataEnvelope(getNBytesWithValues(1000, (n, index) -> index < n % 2 ? (byte) 0 : (byte) 1)) },
+                { new TestDataEnvelope(randomBytesFromGeometricDistribution(1000, 0.1)) },
+                { new TestDataEnvelope(randomBytesFromGeometricDistribution(1000, 0.01)) },
+                { new TestDataEnvelope(randomBytesFromGeometricDistribution(10 * 1000 * 1000 + 1, 0.01)) },
+        };
+    }
+
+    public Object[][] getRansTestDataTinySmallLarge() {
+        return new Object[][]{
+                // params: test data, lower limit, upper limit
+                { new TestDataEnvelope(randomBytesFromGeometricDistribution(100, 0.1)), 1, 100 }, // Tiny
+                { new TestDataEnvelope(randomBytesFromGeometricDistribution(1000, 0.01)), 4, 1000 }, // Small
+                { new TestDataEnvelope(randomBytesFromGeometricDistribution(100 * 1000 + 3, 0.01)), 100 * 1000 + 3 - 4, 100 * 1000 + 3 } // Large
         };
     }
 
-    @Test(dataProvider="ransData")
-    public void testRANS(final TestCaseWrapper tc) {
-        roundTripForEachOrder(tc.testArray);
+    @DataProvider(name="rans4x8")
+    public Object[][] getRans4x8Codecs() {
+        final RANS4x8Encode rans4x8Encode = new RANS4x8Encode();
+        final RANS4x8Decode rans4x8Decode = new RANS4x8Decode();
+        return new Object[][]{
+                {rans4x8Encode, rans4x8Decode, new RANS4x8Params(RANSParams.ORDER.ZERO)}, // RANS4x8 Order 0
+                {rans4x8Encode, rans4x8Decode, new RANS4x8Params(RANSParams.ORDER.ONE)} // RANS4x8 Order 1
+        };
     }
 
-    @Test
-    public void testSizeRangeTiny() {
-        for (int i = 0; i < 20; i++) {
-            final byte[] data = randomBytesFromGeometricDistribution(100, 0.1);
-            final ByteBuffer in = ByteBuffer.wrap(data);
-            for (int size = 1; size < data.length; size++) {
-                in.position(0);
-                in.limit(size);
-                roundTripForEachOrder(in);
-            }
-        }
+    @DataProvider(name="ransNx16")
+    public Object[][] getRansNx16Codecs() {
+        final RANSNx16Encode ransNx16Encode = new RANSNx16Encode();
+        final RANSNx16Decode ransNx16Decode = new RANSNx16Decode();
+        // TODO: More formatFlags values i.e, combinations of bit flags will be added later
+        return new Object[][]{
+
+                //RANSNx16 formatFlags(first byte) 0: Order 0, N = 4, CAT false
+                {ransNx16Encode, ransNx16Decode, new RANSNx16Params(0x00)} ,
+
+                //RANSNx16 formatFlags(first byte) 1: Order 1, N = 4, CAT false
+                {ransNx16Encode, ransNx16Decode, new RANSNx16Params(0x01)} ,
```

Comment:
Does this apply only to hex literals? These different hex values in the data provider `ransNx16` are just random combinations of format flags for test cases.

Reply:
Right, but rather than using comments to explaining what hex values mean, it would be better to just have code that is self-documenting, and uses the real flag values, i.e.:

`{ransNx16Encode, ransNx16Decode, new RANSNx16Params(RANSNx16Params.ORDER_FLAG_MASK)},` 

or, for a more interesting example:

`{ransNx16Encode, ransNx16Decode, new RANSNx16Params(RANSNx16Params.ORDER_FLAG_MASK | RANSNx16Params.RLE_FLAG_MASK)},`

Reply:
Thanks for the example. I understand it now.

Reply:
done